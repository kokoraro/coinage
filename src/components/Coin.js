import React from 'react';

const Coin = ({ coinData, currDollar, addOpen, handleDelete }) => {
  const currency = currDollar ? 'USD' : 'GBP';
  const currSymbol = currDollar ? '$' : 'Â£';
  const value = Number(coinData.price[currency]).toFixed(2);
  const price = `${currSymbol}${value}`;

  const firstDiv = addOpen ? (
    <div className="coin-delete" onClick={() => handleDelete(coinData.id)} >
      <i className="fa fa-trash"></i>
    </div>
  ) : (
    <div className="coin-id">{coinData.id}</div>
  ) 

  return (
    <li className="coin">
      {firstDiv}
      <div className="coin-code">{coinData.code}</div>
      <div className="coin-name">{coinData.name}</div>
      <div className="coin-price">{price}</div>
    </li>
  )
}

export default Coin;