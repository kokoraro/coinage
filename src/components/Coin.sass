.coin img
  width: 25px
  margin: 0 auto

.coin-delete
  color: red
  font-size: 1.3rem

.coin-name
  grid-column: 3 / 6

.coin-price
  text-align: right
  padding-right: 10px
  grid-column: 6 / 8