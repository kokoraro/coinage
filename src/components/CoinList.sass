.coin-list
  list-style: none

  li
    display: grid
    grid-template-columns: repeat(7, 1fr)
    padding: 25px 0
    border-top: 2px solid lightgrey
    position: relative

    &:last-of-type
      border-bottom: 2px solid lightgrey

    &:hover .coin-delete
      display: block

ul
  padding: 0