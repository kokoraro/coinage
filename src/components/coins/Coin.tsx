import React from 'react';
import { Coin } from '../../redux/coins/types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface CoinProps {
    coinData: Coin;
    currDollar: boolean;
    isFavouritesView: boolean
}

const formatCoinPrice = (coinData: Coin, currDollar: boolean) => {
    const currency = currDollar ? 'USD' : 'GBP';
    const currSymbol = currDollar ? '$' : 'Â£';
    const price = Number(coinData.price[currency]);
    const value = price >= 1 ? price.toFixed(2) : price.toFixed(4);

    return `${currSymbol}${value}`;
}


const CoinItem = ({
    coinData,
    currDollar,
    isFavouritesView,
}: CoinProps) => {


    return (
        <div className="coin">
            <img src={coinData.imageURL && `https://www.cryptocompare.com${coinData.imageURL}`} alt={coinData.name} />
            <div className="coin-code">{coinData.code}</div>
            <div className="coin-name">{coinData.name}</div>
            {isFavouritesView ? (
                <FontAwesomeIcon icon="star" />
            ) : (
                <div className="coin-price">{formatCoinPrice(coinData, currDollar)}</div>
            )}
        </div>
    );
};

export default CoinItem;
