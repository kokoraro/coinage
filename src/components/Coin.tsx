import React from 'react';
import './Coin.sass';

interface CoinProps {
    coinData: any;
    currDollar: boolean;
    addOpen: boolean;
    handleDelete: any;
}

const Coin = ({
    coinData,
    currDollar,
    addOpen,
    handleDelete,
}: CoinProps) => {
    const currency = currDollar ? 'USD' : 'GBP';
    const currSymbol = currDollar ? '$' : 'Â£';
    const price = Number(coinData.price[currency]);
    const value = price >= 1 ? price.toFixed(2) : price.toFixed(4);

    return (
        <li className="coin">
            {addOpen ? (
                <div
                    role="button"
                    tabIndex={0}
                    className="coin-delete"
                    onClick={() => handleDelete(coinData.id)}
                >
                    <i className="fas fa-trash-alt" />
                </div>
            ) : (
                <img src={coinData.imageURL && `https://www.cryptocompare.com${coinData.imageURL}`} alt={coinData.name} />
            )}
            <div className="coin-code">{coinData.code}</div>
            <div className="coin-name">{coinData.name}</div>
            <div className="coin-price">{`${currSymbol}${value}`}</div>
        </li>
    );
};

export default Coin;
