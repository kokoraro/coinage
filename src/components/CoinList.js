import React from 'react';
import Coin from './Coin';

const CoinList = ({ coinData, currDollar }) => {
  const coins = coinData ? (
    coinData.map(c => (
      <Coin coinData={c} currDollar={currDollar} key={c.id} />
    ))
  ) : null;
  return (
    <ul className="coin-list">
      {coins} 
    </ul>
  )
}

export default CoinList;