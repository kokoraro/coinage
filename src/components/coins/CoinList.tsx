import React from 'react';
import CoinListItem from './CoinListItem';
import LoadingPanel from '../ui/LoadingPanel';
import { useQuery } from 'react-query';
import { Coin, fetchCoinPrice } from '../api/coins';
import { useGlobalStateContext } from '../../utils/GlobalStateProvider';

interface CoinListProps {
    coinData: Coin[];
    loading: boolean;
}

type CoinPriceResponse = { [key: string]: { GBP: string; USD: string } };

const DEFAULT_COIN_PRICE = { GBP: '0', USD: '0' };

const CoinList = ({
    coinData,
    loading,
}: CoinListProps) => {
    const { state } = useGlobalStateContext();
    const { activeCoinCodes } = state;
    const { isLoading: isLoadingPrice, data: priceData } = useQuery<CoinPriceResponse, Error>('coinsPrice', () => fetchCoinPrice(activeCoinCodes), { refetchOnMount: false });

    if (loading || isLoadingPrice) return <LoadingPanel />;

    const favouriteCoins = coinData.filter(coin => activeCoinCodes.includes(coin.code));

    return (
        <div className="coin-list">
            {favouriteCoins.map(coin => (
                <CoinListItem
                    coin={coin}
                    isCurrencyDollar={state.isCurrencyDollar}
                    key={coin.id}
                    coinPrice={priceData?.[coin.code] || DEFAULT_COIN_PRICE}
                />
            ))}
        </div>
    );
}

export default CoinList;
