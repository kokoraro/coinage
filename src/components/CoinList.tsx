import React from 'react';
import CoinItem from './Coin';
import '../styles/CoinList.sass';
import { Coin } from '../reducers/coins';

interface CoinList {
    coinData: Coin[];
    currDollar: boolean;
    addOpen: boolean;
    handleDelete: (id: number) => void;
}

const CoinList = ({
    coinData,
    currDollar,
    addOpen,
    handleDelete,
}: CoinList) => (
    <ul className="coin-list">
        {coinData.map(c => (
            <CoinItem
                coinData={c}
                currDollar={currDollar}
                key={c.id}
                addOpen={addOpen}
                handleDelete={handleDelete}
            />
        ))}
    </ul>
);

export default CoinList;
