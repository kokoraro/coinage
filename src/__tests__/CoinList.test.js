import React from 'react';
import { shallow } from 'enzyme';
import CoinList from '../components/CoinList';
import Coin from '../components/Coin'

const app = shallow(<CoinList />);

describe("Coin list component", () => {
  it("renders correctly", () => {
    expect(app).toMatchSnapshot();
  });

  it("shows a coin component", () => {
    expect(app.find(Coin).length).toBe(1);
  });
});