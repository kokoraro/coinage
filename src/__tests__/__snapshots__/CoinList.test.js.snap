// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin list component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CoinList
    coinData={
      Array [
        Object {
          "id": 1,
          "name": "Bitcoin",
          "price": 2000,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Coin
          coinData={
            Object {
              "id": 1,
              "name": "Bitcoin",
              "price": 2000,
            }
          }
          currDollar={undefined}
        />,
      ],
      "className": "coin-list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "coinData": Object {
            "id": 1,
            "name": "Bitcoin",
            "price": 2000,
          },
          "currDollar": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Coin
            coinData={
              Object {
                "id": 1,
                "name": "Bitcoin",
                "price": 2000,
              }
            }
            currDollar={undefined}
          />,
        ],
        "className": "coin-list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "coinData": Object {
              "id": 1,
              "name": "Bitcoin",
              "price": 2000,
            },
            "currDollar": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
